File: Sources/ActivityTracker/Tracking/MeetingDetector.swift
import Foundation
import AppKit
import ApplicationServices

/// Detects an in-progress meeting in a supported app (currently Microsoft Teams),
/// independent of whether that app is frontmost — so a Teams call keeps accruing
/// its own time even while you're actively working in a different app (e.g. Chrome
/// on a second monitor). Logs to `meeting_sessions`, a separate table from
/// `app_intervals` specifically because it's allowed to overlap.
///
/// Heuristic (best-effort — there's no public "am I in a call" API for a
/// third-party local tool to query):
///
/// Derived from watching real Teams windows across three states: not in a meeting,
/// a scheduled channel meeting, and an ad-hoc 1:1 call started from a chat. Window
/// *title* wording varies a lot between those ("Meeting in <channel>" vs. just the
/// other person's name, with no "meeting" wording at all) — but structurally, in
/// every meeting state Teams opens a *second* window distinct from the normal
/// single "Chat | ..." browsing window. So: 2+ windows, at least one NOT titled
/// like a "Chat | ..." browsing view, is treated as "in a meeting".
///
/// Known false-positive risk: popping a chat out into its own separate window
/// (a real Teams feature) without being in a call would also produce 2+ windows
/// and could be mistaken for a meeting. Flagged here for future refinement rather
/// than silently accepted — if this turns out to misfire in practice, the fix is
/// to also require a positive signal (e.g. a window title containing "Meeting",
/// which was present in both meeting states we captured, just not guaranteed for
/// every meeting type).
final class MeetingDetector {
    private struct SupportedApp {
        let bundleIDs: [String]
        let displayName: String
    }

    private let supportedApps = [
        SupportedApp(bundleIDs: ["com.microsoft.teams2", "com.microsoft.teams"], displayName: "Microsoft Teams")
    ]

    private var timer: DispatchSourceTimer?
    private var inMeeting: Set<String> = [] // display names currently considered "in a meeting"

    private let pollInterval: TimeInterval = 10.0

    func start() {
        guard timer == nil else { return }
        let t = DispatchSource.makeTimerSource(queue: .main)
        t.schedule(deadline: .now() + 3, repeating: pollInterval)
        t.setEventHandler { [weak self] in self?.poll() }
        t.resume()
        timer = t
        Log.info("MeetingDetector started")
    }

    func stop() {
        timer?.cancel()
        timer = nil
    }

    private func poll() {
        guard AXIsProcessTrusted() else { return } // same permission WindowTitleTracker needs; degrade silently

        for app in supportedApps {
            let running = NSWorkspace.shared.runningApplications.first {
                guard let id = $0.bundleIdentifier else { return false }
                return app.bundleIDs.contains(id)
            }

            guard let running else {
                // App isn't even running — make sure we don't leave a session open.
                if inMeeting.contains(app.displayName) {
                    inMeeting.remove(app.displayName)
                    ActivityStore.shared.endOpenMeetingSession(appName: app.displayName)
                    Log.info("Meeting ended (app quit): \(app.displayName)")
                }
                continue
            }

            let titles = windowTitles(pid: running.processIdentifier)
            let nowInMeeting = Self.looksLikeMeeting(windowTitles: titles)
            let wasInMeeting = inMeeting.contains(app.displayName)

            if nowInMeeting, !wasInMeeting {
                inMeeting.insert(app.displayName)
                ActivityStore.shared.startMeetingSessionIfNeeded(appName: app.displayName)
                Log.info("Meeting detected: \(app.displayName)")
            } else if !nowInMeeting, wasInMeeting {
                inMeeting.remove(app.displayName)
                ActivityStore.shared.endOpenMeetingSession(appName: app.displayName)
                Log.info("Meeting ended: \(app.displayName)")
            }
        }
    }

    static func looksLikeMeeting(windowTitles: [String]) -> Bool {
        guard windowTitles.count >= 2 else { return false }
        return windowTitles.contains { !$0.lowercased().hasPrefix("chat |") }
    }

    private func windowTitles(pid: pid_t) -> [String] {
        let appElement = AXUIElementCreateApplication(pid)
        var windowsRef: AnyObject?
        guard AXUIElementCopyAttributeValue(appElement, kAXWindowsAttribute as CFString, &windowsRef) == .success,
              let windows = windowsRef as? [AXUIElement] else { return [] }

        return windows.compactMap { window in
            var titleRef: AnyObject?
            guard AXUIElementCopyAttributeValue(window, kAXTitleAttribute as CFString, &titleRef) == .success else { return nil }
            return titleRef as? String
        }
    }

    deinit { stop() }
}

File: windows/src/ActivityTracker.Core/Tracking/MeetingDetector.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text;
using System.Threading;
using ActivityTracker.Core.Database;

namespace ActivityTracker.Core.Tracking;

/// <summary>
/// Detects an in-progress Microsoft Teams meeting, independent of whether
/// Teams is the foreground app — the Windows port of the macOS app's
/// MeetingDetector. Logs to meeting_sessions, a separate table from
/// app_intervals specifically because it's allowed to overlap (a call can run
/// in the background while a different app is the one actually tracked).
///
/// UNVERIFIED HEURISTIC — this is the one piece of this port I could not
/// validate myself (no Windows machine here). The macOS version's working
/// heuristic came from watching real Teams windows across three states (not
/// in a meeting / channel meeting / 1:1 call) and finding a structural pattern:
/// every meeting state opened a second window distinct from the normal single
/// "Chat | ..." browsing window — NOT from guessing at title wording, which
/// varied a lot between meeting types. This port assumes the same structural
/// pattern likely holds on the Windows Teams client (it's the same app,
/// probably similar window architecture), but that's an assumption, not a
/// confirmed fact — Teams' Windows and macOS clients aren't guaranteed to
/// behave identically. IsLikelyInMeeting logs full window counts/titles every
/// poll specifically so this can be validated/corrected the same way the
/// macOS heuristic was: watch the real log during an actual call and adjust
/// LooksLikeMeeting from what it actually shows.
/// </summary>
public sealed class MeetingDetector
{
    private static readonly string[] TeamsProcessNames = ["ms-teams", "Teams"];
    private const string DisplayName = "Microsoft Teams";

    private Timer? _timer;
    private bool _inMeeting;
    private static readonly TimeSpan PollInterval = TimeSpan.FromSeconds(10);

    public void Start()
    {
        if (!OperatingSystem.IsWindows())
        {
            Log.Info("MeetingDetector: not running on Windows, skipping");
            return;
        }
        if (_timer != null) return;
        _timer = new Timer(_ => Poll(), null, TimeSpan.FromSeconds(3), PollInterval);
        Log.Info("MeetingDetector started");
    }

    public void Stop()
    {
        _timer?.Dispose();
        _timer = null;
    }

    private void Poll()
    {
        Process[]? teamsProcesses = null;
        foreach (var name in TeamsProcessNames)
        {
            var found = Process.GetProcessesByName(name);
            if (found.Length > 0) { teamsProcesses = found; break; }
        }

        if (teamsProcesses is null || teamsProcesses.Length == 0)
        {
            if (_inMeeting)
            {
                _inMeeting = false;
                ActivityStore.Shared.EndOpenMeetingSession(DisplayName);
                Log.Info("Meeting ended (Teams not running)");
            }
            return;
        }

        var pids = new HashSet<int>();
        foreach (var p in teamsProcesses) pids.Add(p.Id);

        var titles = EnumerateVisibleWindowTitles(pids);
        // Diagnostic logging every poll — see the class doc. This is the data
        // that needs collecting from a real call to confirm or correct the
        // heuristic below.
        Log.Info($"MeetingDetector: Teams has {titles.Count} visible window(s): [{string.Join(" | ", titles)}]");

        var nowInMeeting = LooksLikeMeeting(titles);
        if (nowInMeeting && !_inMeeting)
        {
            _inMeeting = true;
            ActivityStore.Shared.StartMeetingSessionIfNeeded(DisplayName);
            Log.Info("Meeting detected: Microsoft Teams");
        }
        else if (!nowInMeeting && _inMeeting)
        {
            _inMeeting = false;
            ActivityStore.Shared.EndOpenMeetingSession(DisplayName);
            Log.Info("Meeting ended: Microsoft Teams");
        }
    }

    /// <summary>
    /// Same structural principle as the macOS version: 2+ windows, at least
    /// one not looking like the normal single chat-browsing view, is treated
    /// as a meeting. "Chat" is a guess at what a normal Teams-for-Windows
    /// browsing window's title looks like — unconfirmed; check the diagnostic
    /// log from a real call and adjust this if it doesn't match reality.
    /// </summary>
    public static bool LooksLikeMeeting(IReadOnlyList<string> windowTitles)
    {
        if (windowTitles.Count < 2) return false;
        foreach (var title in windowTitles)
        {
            if (!title.Contains("Chat", StringComparison.OrdinalIgnoreCase)) return true;
        }
        return false;
    }

    private static List<string> EnumerateVisibleWindowTitles(HashSet<int> pids)
    {
        var titles = new List<string>();
        Win32.EnumWindows((hWnd, _) =>
        {
            if (!Win32.IsWindowVisible(hWnd)) return true;
            Win32.GetWindowThreadProcessId(hWnd, out var pid);
            if (!pids.Contains((int)pid)) return true;

            var sb = new StringBuilder(512);
            var length = Win32.GetWindowText(hWnd, sb, sb.Capacity);
            if (length > 0) titles.Add(sb.ToString());
            return true;
        }, IntPtr.Zero);
        return titles;
    }
}
